File: frontend/src/api/workflow.ts
import { request } from '@/utils/request'
import type { ApiResponse, PageData, PageParams } from '@/types/api'

// 流程定义类型
export interface WorkflowDefinition {
  id: number
  name: string
  code: string
  type: string
  version: string
  config: string
  status: string
  description: string
  created_at: string
  updated_at: string
  created_by: number
}

// 流程实例类型
export interface WorkflowInstance {
  id: number
  definition_id: number
  business_type: string
  business_id: number
  title: string
  status: string
  initiator_id: number
  current_node: string
  comment: string
  created_at: string
  updated_at: string
  definition?: WorkflowDefinition
  initiator?: {
    id: number
    username: string
    real_name: string
  }
}

// 流程实例响应（含业务对象信息）
export interface WorkflowInstanceResponse extends WorkflowInstance {
  business_name: string
  business_code: string
  business_status: string
}

// 流程历史类型
export interface WorkflowHistory {
  id: number
  instance_id: number
  node_name: string
  action: string
  operator_id: number
  operator_name: string
  comment: string
  created_at: string
  operator?: {
    id: number
    username: string
    real_name: string
  }
}

// 待办事项响应
export interface TodoItemResponse extends WorkflowInstanceResponse {
  pending_action: string
}

// 流程定义列表查询参数
export interface DefinitionListParams extends PageParams {
  type?: string
  status?: string
  keyword?: string
}

// 流程实例列表查询参数
export interface InstanceListParams extends PageParams {
  type?: string
  status?: string
  keyword?: string
}

// 待办列表查询参数
export interface TodoListParams extends PageParams {
  type?: string    // 业务类型筛选
  status?: string
}

// 创建流程定义请求
export interface CreateDefinitionRequest {
  name: string
  code: string
  type: string
  config: string
  description?: string
}

// 更新流程定义请求
export interface UpdateDefinitionRequest {
  name?: string
  config?: string
  description?: string
}

// 发起流程请求
export interface InitiateWorkflowRequest {
  definition_id: number
  business_type: string
  business_id: number
  title: string
  comment?: string
}

// 同意审批请求
export interface ApproveRequest {
  comment?: string
}

// 驳回审批请求
export interface RejectRequest {
  comment: string
}

// 转交审批请求
export interface TransferRequest {
  target_user_id: number
  comment?: string
}

// ==================== 流程定义API ====================

// 获取流程定义列表
export function getDefinitionList(params: DefinitionListParams): Promise<ApiResponse<PageData<WorkflowDefinition>>> {
  return request.get('/workflow/workflows/definitions', { params })
}

// 获取流程定义详情
export function getDefinition(id: number): Promise<ApiResponse<WorkflowDefinition>> {
  return request.get(`/workflow/workflows/definitions/${id}`)
}

// 创建流程定义
export function createDefinition(data: CreateDefinitionRequest): Promise<ApiResponse<WorkflowDefinition>> {
  return request.post('/workflow/workflows/definitions', data)
}

// 更新流程定义
export function updateDefinition(id: number, data: UpdateDefinitionRequest): Promise<ApiResponse<WorkflowDefinition>> {
  return request.put(`/workflow/workflows/definitions/${id}`, data)
}

// 发布流程定义
export function releaseDefinition(id: number): Promise<ApiResponse<WorkflowDefinition>> {
  return request.post(`/workflow/workflows/definitions/${id}/release`)
}

// 删除流程定义
export function deleteDefinition(id: number): Promise<ApiResponse<null>> {
  return request.delete(`/workflow/workflows/definitions/${id}`)
}

// 获取指定类型的激活流程定义
export function getActiveDefinition(type: string): Promise<ApiResponse<WorkflowDefinition>> {
  return request.get('/workflow/workflows/definitions/active', { params: { type } })
}

// ==================== 流程实例API ====================

// 获取流程实例列表
export function getInstanceList(params: InstanceListParams): Promise<ApiResponse<PageData<WorkflowInstanceResponse>>> {
  return request.get('/workflow/workflows/instances', { params })
}

// 获取流程实例详情
export function getInstance(id: number): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.get(`/workflow/workflows/instances/${id}`)
}

// 发起流程
export function initiateWorkflow(data: InitiateWorkflowRequest): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.post('/workflow/workflows/instances', data)
}

// 获取流程历史
export function getHistories(instanceId: number): Promise<ApiResponse<WorkflowHistory[]>> {
  return request.get(`/workflow/workflows/instances/${instanceId}/history`)
}

// ==================== 待办API ====================

// 获取我的待办
export function getMyTodos(params: TodoListParams): Promise<ApiResponse<PageData<TodoItemResponse>>> {
  return request.get('/workflow/workflows/todos', { params })
}

// 获取待办数量
export function getTodoCount(): Promise<ApiResponse<{ count: number }>> {
  return request.get('/workflow/workflows/todos/count')
}

// ==================== 审批操作API ====================

// 同意审批
export function approveWorkflow(instanceId: number, data: ApproveRequest): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.post(`/workflow/workflows/instances/${instanceId}/approve`, data)
}

// 驳回审批
export function rejectWorkflow(instanceId: number, data: RejectRequest): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.post(`/workflow/workflows/instances/${instanceId}/reject`, data)
}

// 转交审批
export function transferWorkflow(instanceId: number, data: TransferRequest): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.post(`/workflow/workflows/instances/${instanceId}/transfer`, data)
}

// 撤回流程
export function withdrawWorkflow(instanceId: number): Promise<ApiResponse<WorkflowInstanceResponse>> {
  return request.post(`/workflow/workflows/instances/${instanceId}/withdraw`)
}

File: backend/internal/modules/workflow/model/workflow.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// WorkflowDefinition 流程定义模型
type WorkflowDefinition struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	Name        string         `gorm:"size:200;not null" json:"name"`
	Code        string         `gorm:"size:50;not null" json:"code"`
	Type        string         `gorm:"size:50;not null;index" json:"type"` // BOM_APPROVAL, DOC_APPROVAL, MATERIAL_APPROVAL
	Version     string         `gorm:"size:2;not null;default:AA" json:"version"`
	Config      string         `gorm:"type:json;not null" json:"config"` // 流程配置JSON
	Status      string         `gorm:"size:20;default:DRAFT;index" json:"status"` // DRAFT, RELEASED, OBSOLETE
	Description string         `json:"description"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	CreatedBy   uint           `gorm:"not null" json:"created_by"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

// WorkflowInstance 流程实例模型
type WorkflowInstance struct {
	ID            uint           `gorm:"primaryKey" json:"id"`
	DefinitionID  uint           `gorm:"not null;index" json:"definition_id"`
	BusinessType  string         `gorm:"size:50;not null;index" json:"business_type"` // MATERIAL, DOCUMENT, BOM
	BusinessID    uint           `gorm:"not null;index:idx_business" json:"business_id"`
	Title         string         `gorm:"size:200" json:"title"`
	Status        string         `gorm:"size:20;default:DRAFT;index" json:"status"` // DRAFT, PENDING, APPROVED, REJECTED, CANCELLED
	InitiatorID   uint           `gorm:"not null;index" json:"initiator_id"`
	CurrentNode   string         `gorm:"size:100" json:"current_node"`                     // 当前节点名称
	CurrentNodeID string         `gorm:"size:50" json:"current_node_id"`                   // 当前节点ID
	NodePath      string         `gorm:"type:text" json:"node_path"`                       // 经过的节点路径（JSON数组）
	ApprovedNodes string         `gorm:"type:text" json:"approved_nodes"`                  // 已审批节点列表（JSON数组）
	Comment       string         `json:"comment"`                                          // 发起时的意见
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`

	// 关联
	Definition *WorkflowDefinition `gorm:"foreignKey:DefinitionID" json:"definition,omitempty"`
	Initiator  *User               `gorm:"foreignKey:InitiatorID" json:"initiator,omitempty"`
}

// WorkflowHistory 流程历史模型
type WorkflowHistory struct {
	ID           uint      `gorm:"primaryKey" json:"id"`
	InstanceID   uint      `gorm:"not null;index" json:"instance_id"`
	NodeName     string    `gorm:"size:100" json:"node_name"`
	FromNodeID   string    `gorm:"size:50" json:"from_node_id"`   // 源节点ID
	ToNodeID     string    `gorm:"size:50" json:"to_node_id"`     // 目标节点ID
	Action       string    `gorm:"size:20;not null" json:"action"` // SUBMIT, APPROVE, REJECT, TRANSFER, CANCEL, WITHDRAW
	OperatorID   uint      `gorm:"not null;index" json:"operator_id"`
	OperatorName string    `gorm:"size:50" json:"operator_name"`
	Comment      string    `json:"comment"`
	CreatedAt    time.Time `json:"created_at"`

	// 关联
	Operator *User `gorm:"foreignKey:OperatorID" json:"operator,omitempty"`
}

// User 简化的用户模型（用于关联查询）
type User struct {
	ID       uint   `gorm:"primaryKey" json:"id"`
	Username string `gorm:"size:50;not null" json:"username"`
	RealName string `gorm:"size:50" json:"real_name"`
}

// CreateWorkflowDefinitionRequest 创建流程定义请求
type CreateWorkflowDefinitionRequest struct {
	Name        string `json:"name" binding:"required,max=200"`
	Code        string `json:"code" binding:"required,max=50"`
	Type        string `json:"type" binding:"required,oneof=MATERIAL_APPROVAL DOCUMENT_APPROVAL BOM_APPROVAL"`
	Config      string `json:"config" binding:"required"`
	Description string `json:"description"`
}

// UpdateWorkflowDefinitionRequest 更新流程定义请求
type UpdateWorkflowDefinitionRequest struct {
	Name        string `json:"name" binding:"max=200"`
	Config      string `json:"config"`
	Description string `json:"description"`
}

// InitiateWorkflowRequest 发起流程请求
type InitiateWorkflowRequest struct {
	DefinitionID uint   `json:"definition_id" binding:"required"`
	BusinessType string `json:"business_type" binding:"required,oneof=MATERIAL DOCUMENT BOM"`
	BusinessID   uint   `json:"business_id" binding:"required"`
	Title        string `json:"title" binding:"required,max=200"`
	Comment      string `json:"comment"`
}

// ApproveRequest 同意审批请求
type ApproveRequest struct {
	Comment string `json:"comment"`
}

// RejectRequest 驳回审批请求
type RejectRequest struct {
	Comment string `json:"comment" binding:"required"`
}

// TransferRequest 转交审批请求
type TransferRequest struct {
	TargetUserID uint   `json:"target_user_id" binding:"required"`
	Comment      string `json:"comment"`
}

// WorkflowListQuery 流程列表查询参数
type WorkflowListQuery struct {
	Page     int    `form:"page"`
	PageSize int    `form:"page_size"`
	Type     string `form:"type"`
	Status   string `form:"status"`
	Keyword  string `form:"keyword"`
}

// TodoListQuery 待办列表查询参数
type TodoListQuery struct {
	Page     int    `form:"page"`
	PageSize int    `form:"page_size"`
	Type     string `form:"type"`   // 业务类型筛选
	Status   string `form:"status"`
	Keyword  string `form:"keyword"` // 关键字筛选
}

// WorkflowDefinitionListQuery 流程定义列表查询参数
type WorkflowDefinitionListQuery struct {
	Page     int    `form:"page"`
	PageSize int    `form:"page_size"`
	Type     string `form:"type"`
	Status   string `form:"status"`
	Keyword  string `form:"keyword"`
	NodeName string `form:"node_name"` // 按节点名称筛选
}

// WorkflowInstanceResponse 流程实例响应
type WorkflowInstanceResponse struct {
	WorkflowInstance
	BusinessName   string   `json:"business_name"`   // 业务对象名称
	BusinessCode   string   `json:"business_code"`   // 业务对象编码
	BusinessStatus string   `json:"business_status"` // 业务对象状态
	NodePathList   []string `json:"node_path_list"`  // 解析后的节点路径
	ApprovedList   []string `json:"approved_list"`   // 解析后的已审批节点
}

// TodoItemResponse 待办事项响应
type TodoItemResponse struct {
	WorkflowInstanceResponse
	PendingAction string `json:"pending_action"` // 待执行操作描述
}

// TableName 指定表名
func (WorkflowDefinition) TableName() string {
	return "workflow_definitions"
}

func (WorkflowInstance) TableName() string {
	return "workflow_instances"
}

func (WorkflowHistory) TableName() string {
	return "workflow_histories"
}

func (User) TableName() string {
	return "users"
}
